skimage/segmentation/active_contour_model.py

```diff
@@ -0,0 +1,204 @@
+                    sobel(img[:, :, 2])]
+        else:
+            edge = [sobel(img)]
+        for i in xrange(3 if RGB else 1):
+            edge[i][0, :] = edge[i][1, :]
+            edge[i][-1, :] = edge[i][-2, :]
+            edge[i][:, 0] = edge[i][:, 1]
+            edge[i][:, -1] = edge[i][:, -2]
+    else:
+        edge = [0]
+
+    # Superimpose intensity and edge images:
+    if RGB:
+        img = w_line*np.sum(img, axis=2) \
+            + w_edge*sum(edge)
+    else:
+        img = w_line*img + w_edge*edge[0]
+
+    # Interpolate for smoothness:
+    intp = RectBivariateSpline(np.arange(img.shape[1]),
+            np.arange(img.shape[0]), img.T, kx=2, ky=2, s=0)
+
+    x, y = snake[:, 0].copy(), snake[:, 1].copy()
+    xsave = np.empty((convergence_order, len(x)))
+    ysave = np.empty((convergence_order, len(x)))
+
+    # Build snake shape matrix
+    n = len(x)
+    a = np.roll(np.eye(n), -1, axis=0) \
+      + np.roll(np.eye(n), -1, axis=1) \
+      - 2*np.eye(n)
+    b = np.roll(np.eye(n), -2, axis=0) \
+      + np.roll(np.eye(n), -2, axis=1) \
+      - 4*np.roll(np.eye(n), -1, axis=0) \
+      - 4*np.roll(np.eye(n), -1, axis=1) \
+      + 6*np.eye(n)
+    A = -alpha*a + beta*b
+
+    # Impose boundary conditions different from periodic:
+    sfixed = False
+    if bc.startswith('fixed'):
+        A[0, :] = 0
+        A[1, :] = 0
+        A[1, :3] = [1, -2, 1]
+        sfixed = True
+    efixed = False
+    if bc.endswith('fixed'):
+        A[-1, :] = 0
+        A[-2, :] = 0
+        A[-2, -3:] = [1, -2, 1]
+        efixed = True
+    sfree = False
+    if bc.startswith('free'):
+        A[0, :] = 0
+        A[0, :3] = [1, -2, 1]
+        A[1, :] = 0
+        A[1, :4] = [-1, 3, -3, 1]
+        sfree = True
+    efree = False
+    if bc.endswith('free'):
+        A[-1, :] = 0
+        A[-1, -3:] = [1, -2, 1]
+        A[-2, :] = 0
+        A[-2, -4:] = [-1, 3, -3, 1]
+        efree = True
+
+    # Only one inversion is needed:
+    inv = scipy.linalg.inv(A+gamma*np.eye(n))
+
+    # Explcit time stepping for image energy minimization:
+    for i in xrange(max_iterations):
+        fx = intp(x, y, dx=1, grid=False)
+        fy = intp(x, y, dy=1, grid=False)
+        if sfixed:
+            fx[0] = 0
+            fy[0] = 0
+        if efixed:
+            fx[-1] = 0
+            fy[-1] = 0
+        if sfree:
+            fx[0] *= 2
+            fy[0] *= 2
+        if efree:
+            fx[-1] *= 2
+            fy[-1] *= 2
+        xn = np.dot(inv, gamma*x + fx)
+        yn = np.dot(inv, gamma*y + fy)
+
+        # Movements are capped to max_px_move per iteration:
+        dx = max_px_move*np.tanh(xn-x)
+        dy = max_px_move*np.tanh(yn-y)
+        if sfixed:
+            dx[0] = 0
+            dy[0] = 0
+        if efixed:
+            dx[-1] = 0
+            dy[-1] = 0
+        x[:] += dx
+        y[:] += dy
+
+        # Convergence criteria:
+        j = i%(convergence_order+1)
```

Comment:
I'm not sure I understand the convergence criterion. Do we have to compare the current value to the 10 previous ones because we might expect oscillations? Why a value of 10 for ``convergence_order``?